### PR TITLE
First draft of Xilem blog post

### LINKED ISSUE #70: Writeup of Rust UI architecture ideas
I have another round of ideas for UI architecture in Rust. It's in the https://github.com/linebender/druid/tree/idiopath/idiopath branch, superseding [lasagna](https://github.com/linebender/druid/blob/lasagna/lasagna/src/rusty.rs). (Note: links to github branches aren't great, probably a PR, even a draft PR is better.)

This is a companion post in a way to #68 but the scope is different. While that one is about the challenge of actually shipping a UI, this one is effectively an academic exploration into the best way to express reactive UI. It's also a followup to previous SF meetup talks and architecture blog posts.

Here's something of an outline:

* Introduction of the problem
  - not sure how much to talk about existing Druid ("muggle") architecture, as most readers probably won't be familiar
  - ideas are based on a study of many UI systems, both in Rust and outside. Elm architecture as a foundation, SwiftUI is probably the closest precedent.
  - constraints: high performance, idiomatic Rust, good ergonomics
* High level overview of the architecture
  - three trees: view, state, widgets
  - app produces view tree, with "fairly short" lifetime
  - reconciliation (diffing) causes update to widget tree (also state tree, which is in support of reconciliation)
  - all 3 trees are strongly typed
  - every node has an id
    + actually messy question: id of node in the view tree, or every node in the widget tree? Most of the time they correspond
  - the widget tree handles mechanics of UI: drawing, processing input, etc
  - it produces (higher level) events, which are sent back to the app
  - key innovation: every event contains the id path (sequence of ids on the path from the root to the node that generated the event)
  - events are dispatched based on this id path
  - view tree contains callbacks, which have mutable access to app state
    + contrast elm, with manual dispatch from root of app
    + contrast nearly every other reactive UI framework, with shared mutable state of some kind
* Deeper dive: components
  - vanilla code for passing state (similar to React props) to child components
  - use of Adapt node to "lens" mutable state access
* Deeper dive: incrementality
  - Memoize node, avoids actually constructing view subtree
  - pointer equality on Rc values
    + in combination with Adapt, is actually sophisticated change tracking
  - ideas should generalize to collections, esp with immutable data types
* Deeper dive: type erasure
  - pattern works even with associated types and Self param (non object safe)
  - TODO: best references for this pattern? Is it known?
* Deeper dive: layering
  - widget tree works by itself, just not ergonomic
  - like Flutter but *unlike* SwiftUI and Elm, where lower layer is off limits
* Topics to touch on
  - accessibility
  - async
  - scripting
* Not in scope: styling
* Acknowledge inspirations and people: many Rust toolkits, Olivier, Manmeet, Taylor, Jan

### PR DESCRIPTION
**Pull request by raphlinus**
Closes #70

### CODE REVIEW

File: _posts/2022-04-24-ui-architecture.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+### Event propagation
+
+Let's click that button. Obviously the goal will be to run that callback and increment the count, but the details of how that happens are subtly different than most declarative UI systems. Probably the "standard" way to do this would be to attach the callback to the button, and have it capture a reference to the chunk of state it mutates. Again, in most declarative systems but not Xilem, setting the new state would be done using some variant of the [observer pattern], for example some kind of `stateState` or other "setter" function to not only update the value but also notify downstream dependencies that it had changed, in this case re-rendering the label.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:89]: 'stateState'? Did you mean 'setState'?

File: _posts/2022-04-24-ui-architecture.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+</details>
+
+So what does Xilem do instead? The view tree is also parameterized on the *app state,* which can be any type. This idea is an evolution of Druid's existing architecture, which also offers mutable access to app state to UI callbacks, but removes some of the limitations. In particular, Druid requires app state to be clonable and diffable, a stumbling block for many new users.
+
+When an event is generated, it is annotated with the path of the UI element that originated it. In the case of the button, `[1, 3]`. We retain the view tree for a fairly short time, but long enough to do event propagation. The event is dispatched from the *root,* and each successive node strips off one id from the id path, choosing a child, and traversing to that child. In this case, that's simple, the enclosing stack just chooses the button. This event propagation also passes down mutable access to the app state.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:103]: This part is quite confusing for a number of reasons:
First, previously to this you showed the IDs being assigned to the widget tree, but here you talk about the view tree. Why do events propagate through views and not widgets?
Secondly, what is meant by 'choosing a child'? How is this choice made? Why in the example is the button chosen over the text?
And thirdly, how does the event propagation provide mutable access to the state?

File: _posts/2022-04-24-ui-architecture.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+Given that setup, running the closure is fairly straightforward. The event is dispatched to the stack, which uses the id to recognize that it needs to be routed to its second child, and that button is the target. It runs the callback on the mutable app state, which in this case just increments `count`.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:105]: Even after reading this I'm still confused how the vstack knows to route the event to the button and how this also provides mutable access. Also, what event is this? Is this something like a mouse button event?

I think this part could do with some more explanation.

File: _posts/2022-04-24-ui-architecture.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+### Re-rendering
+
+After clicking the button and running the callback, the app state consists of the number 1, formerly 0. The app logic function is run, producing a new view tree, and this time the string value is "Count: 1" rather than "Count: 0". The challenge is then to update the widget tree with the new data.
+
+As is completely standard in declarative UI, it is done by diffing the old view tree against the new one, in this case calling the `rebuild` method on the `View` trait. This method compares the data, updates the associated widget if there are any changes, and also traverses into children.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:111]: How does this diffing occur if the view tree is short-lived? Is there a copy of the view tree being stored accross cycles?

File: _posts/2022-04-24-ui-architecture.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+In the simplest case, the app builds the full view tree, and that is diffed in full against the previous version. However, as UI scales, this would be inefficient, so there are *other* mechanisms to do finer grained change propagation, as described below.
+
+## Components
+
+The above is the basic architecture, enough to get started. Now we will go into some more advanced techniques.
+
+It would be very limiting to have a single "app state" type throughout the application, and require all callbacks to express their state mutations in terms of that global type. So we won't do that.
+
+The main tool for stitching together components is the `Adapt` view node. This node is so named because it adapts between one app state type and another, using a closure that takes mutable access to the parent state, and calls into a child (through a "thunk") with a mutable reference to the child state.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:121]: You haven't defined what a 'component' is. Previous to this the reader knows only about views, widgets, and view state. Is a component just another term for a view? This section is titled 'components' but talks mostly about the adapt node (view?).

File: _posts/2022-04-24-ui-architecture.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+In the simple case where the child component operates independently of the parent, the adapt node is a couple lines of code. It is also an attachment point for richer interactions - the closure can manipulate the parent state in any way it likes. The event handler of the child component is also allowed to return an arbitrary type (unit by default), for upward propagation of data.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:123]: I'm not sure I understand what any of this means. Maybe an example would help? Does this section require knowledge of Druid because I'm not sure what it means by 'event handler' and 'return an arbitrary type'.

File: _posts/2022-04-24-ui-architecture.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+Ron Minsky has [stated][Signals and Threads: Building a UI framework] "hidden inside of every UI framework is some kind of incrementalization framework." Xilem unapologetically contains at its core a lightweight change propagation engine, similar in scope to the attribute graph of SwiftUI, but highly specialized to the needs of UI, and in particular with a lightweight approach to *downward* propagation of dependencies, what in React would be stated as the flow of props into components.
+
+In this particular case, that incremental change propagation is best represented as a *memoization* node, yet another implementation of the View trait. A memoization node takes a data value (which supports both `Clone` and equality testing) and a closure which accepts that same data type. On rebuild, it compares the data value with the previous version, and only runs the closure if it has changed. The signature of this node is very similar to [Html.Lazy] in Elm.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:133]: Previously you said that Xilem:
> removes some of the limitations. In particular, Druid requires app state to be clonable and diffable, a stumbling block for many new users

But this seems to be re-introducing that limitation if the user wants this finer grained change propagation.

File: _posts/2022-04-24-ui-architecture.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        thunk.call(&mut child),
+        if !Arc::ptr_eq(&data.child, &child) {
+            Arc::make_mut(data).child = child;
+        }
+    },
+    child_view(...) // which has Arc<Child> as its app data type
+)
+```
+
+This logic propagates the change up the tree *only if* the child state has actually changed.
```
Comment (geom3trik) [inline _posts/2022-04-24-ui-architecture.md:173]: Don't you want change to propagate down the tree? Actually, which tree is this referring to?

**Review by geom3trik — COMMENTED** [review]
Very interesting read overall but a few things came to mind while reading it:

- I think the part detailing the id path stuff could do with a sentence or two explaining why this is preferable to the global state approach you mention.
- The section on event propgation and mutable access to the app state made no sense to me personally. I will check the prototype code on this but I think the explanation could do with being expanded with some more example code.
- Possibly same for the adapt and memoize nodes but if it's just a high level overview you're aiming for then those sections are fine.
- I would be interested to know how this system deals with 'local state' that might be needed by a subtree of widgets but isn't derived from the app state. Is there a special node for this, or something else?

Maybe of-topic for this post but I would be interested to hear your thoughts on state management at some point. When comparing to state management in the elm architecture you stated that:
> Some people like the explicitness of this approach, but it is unquestionably more verbose than a single callback that manipulates state directly as in React or SwiftUI.

It's my understanding that having app state centralized and updated with events actually makes it easier to scale an application, which is why Redux is so popular in React. I can infer that you have a different opinion on that but I would be interested to hear what your thoughts are on the scalability of Xilem's approach.

**Comment by raphlinus** [comment]
Thanks for the detailed commentary @geom3trik! I've been so close to these problems so long it's not easy to get calibrated on what I can take for granted and what needs to be explained in more detail. This will definitely help make the post better, and I hope to have another draft out soon.

File: _posts/2022-04-24-ui-architecture.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+The most innovative aspect of Xilem is event dispatching based on an *id path,* at each stage providing mutable access to app state. A distinctive feature is Adapt nodes (an evolution of the lensing concept in Druid) which facilitate composition of components. By routing events *through* Adapt nodes, subcomponents have access to a different mutable state reference than the parent.
+
+## Synchronized trees
+
+In each "cycle," the app produces a view treeRendering in Xilem begins with the view tree. This tree has fairly short lifetime; each time the UI is updated, a new tree is generated. From this, a widget tree is built (or rebuilt), and the view tree is retained only long enough to assist in event dispatching and then be diffed against the next version, at which point it is dropped. In addition to these two trees, there is a third tree containing *view state,* which persists across cycles. (The view state serves a very similar function as React hooks)
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply _posts/2022-04-24-ui-architecture.md:19]: The view state is stored as a tree, yes. I'm not familiar with Recoil but I certainly am aware of more general incremental computation engines (Salsa, Adapton, Incremental, etc) in which the dependencies are explicitly modeled as a graph.

This is all potentially a much deeper conversation. I believe, without having gone too deep into implementation, that it would be fairly straightforward to wire up an actual graph incremental engine, and have the view state tree consist of lightweight references into that engine. (such an engine could be made available through context/environment, which is not currently part of the implementation but is planned)

So basically this is an area where I want to see what happens.

File: _posts/2022-04-24-ui-architecture.md [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+The Rust expression of these trees is instances of the `View` trait, which has two associated types, one for view state and one for the associated widget. The state and widgets are *also* statically typed. The design relies *heavily* on the type inference mechanisms of the Rust compiler. In addition to inferring the type of the view tree, it also uses associated types to deduce the type of the associated state tree and widget tree, which are known at compile time. In almost every other comparable system (SwiftUI being the notable exception), these are determined at runtime with a fair amount of allocation, downcasting, and dynamic dispatch.
+
+## A worked example
+
+We'll use the classic counter as a running example. It's very simple but will give insight into how things work under the hood. For people who want to follow along with the code, check the idiopath directory of the idiopath branch; running `cargo doc --open` there will reveal a bunch of Rustdoc.
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply _posts/2022-04-24-ui-architecture.md:27]: I also added a link. I'm slightly uncertain about all this as there's a chance I might rename the branch/directory with the "Xilem" branding, but I'll leave it as-is for now.

File: _posts/2022-04-24-ui-architecture.md [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+       | VStack |
+       +--------+
+        /      \
+   2   /      3 \
+   +------+   +--------+
+   | Text |   | Button |
+   +------+   +--------+
+```
+
+The idea of assigning a stable identity to a widget is quite standard in declarative UI (it's also present in basically all non-toy immediate mode GUI implementations), but Xilem adds a distinctive twist, the use of *id path* rather than a single id. The id path of a widget is the sequence of all ids on the path from the root to that widget in the widget tree. Thus, the id path of the button in the above is `[1, 3]`, while the label is `[1, 2]` and the stack is just `[1]`. The full id path is redundant if we had global information about the structure of the tree (for example, by following parent links), but the point is that given id paths, we don't *need* to track this kind of information.
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply _posts/2022-04-24-ui-architecture.md:85]: I added something, but I think you've asked a deep question here, and I'm not positive I can give a good argument at the level of writing I'm doing.

I did indeed in an earlier prototype have tree structure tracking that was done on construction/mutation of the widget tree, so the is visible during reconciliation were just plain integer tokens (not paths). One of the subtle ways that breaks down is ids allocated to futures - they might not correspond to an actual node in the widget tree, or perhaps dummy nodes would be added. Same for env_get, which I've written about on Zulip but not in this blog.

Note also that Elm has an analogous problem; Html has `map` but there is no comparable component-forming operation for subscriptions. As far as I know, those have to be addressed to the top-level update method of the app logic.

I'll think a little more about whether I can explain things better. Most often, when I find things hard to explain, it's because of a gap in my own understanding.

File: _posts/2022-04-24-ui-architecture.md (hunk 7 above)
Comment (raphlinus) [reply _posts/2022-04-24-ui-architecture.md:133]: Yes. It's a tradeoff, and the difference is that it's now opt-in, while Druid absolutely required it.

File: _posts/2022-04-24-ui-architecture.md (hunk 8 above)
Comment (raphlinus) [reply _posts/2022-04-24-ui-architecture.md:173]: The app state tree. But I'm going to change the way I word this, because there's no *requirement* that the app state be modeled as a tree, unlike original Druid. It's quite fine for the app state to be a graph (and we're going to have this in Runebender, as glyphs referencing components is definitely graph rather than tree structure). Good catch, thanks.

File: _posts/2022-04-24-ui-architecture.md (hunk 2 above)
Comment (raphlinus) [reply _posts/2022-04-24-ui-architecture.md:103]: I will rewrite this.

The question of whether ids refer to views or widgets is extremely deep and subtle. In most cases, where views correspond directly to widgets, the *lifetime* of the id should correspond exactly to the lifetime of the widget. In addition, if the construction of the widget tree allows ids to be given on construction (a fine implementation choice), then the widget ids can be exactly the ids allocated during reconciliation. However, as I've pointed out in another comment, it's also fine to allocate ids without a corresponding widget, and it would be possible for the ids here and the widget ids to be disjoint.

At the same time, calling these ids of view nodes feels confusing too, as the view tree has a short lifetime and these ids are stable.

I'll also note, the SwiftUI presentations are maximally confusing around this point, as they consistently fail to make a distinction between a transitory view as generated by the app and what I call an "instantiated view" which is more persistent. The latter is completely invisible and opaque to the *app,* but of course it is easy to observe widget behavior, for example whether animations play continuously.

I think the missing info here is that containers hold the ids of their children in their associated view state. I will definitely add that, as otherwise it would be very mysterious how local routing is done.

**Comment by black7375** [comment]
Disclaimer: I'm a Web front end developer, and not familiar with native UI.

Web frameworks have some interesting ideas.
I can share a few things.

**React's concurrent mode**

I heard that [druid is switching to a tic model](https://github.com/linebender/druid/issues/1945), which I think it is possible to scheduling.

React works in [concurrecy mode](https://en.reactjs.org/docs/concurrent-mode-intro.html) to prevent blocking.

> We’re rendering the results! As soon as React finishes rendering the first slider update, it begins to render the transition to the results. Since this update is opted-into concurrent rendering, React will do three new things:
>
> - Yielding: every 5 ms, React will stop working to allow the browser to do other work, like run promises or fire events. This is the reason why the big chunk of work from the previous examples is now chopped up into smaller pieces. React has split up everything that it needs to do into smaller pieces of work, and is smart enough to pause and let the browser handle pending events (this is called “yielding”). In our case, yielding allows the browser to fire more mousemove events from the Slider to tell React that the mouse is still moving.
> - Interrupting: When a second mouse move event comes in, we schedule another update for the Slider to move it. But how does React render that update, if we’re already in the middle of rendering the results from the last update? The answer is that when React starts working again, it will see that a new urgent update was scheduled during the mouse move and stop working on the pending results (since they’re out of date anyway). React switches to rendering the urgent Slider update, and when that urgent work is done, it goes back to rendering the results. We call this “interrupting” because rendering the results are “interrupted” to render the Slider.
> - Skipping old results: If React just started rendering the first results it was working on, then it would start to build up a queue of results to render, which would take too much time (like the setTimeout example above). So what React does instead is skips the old work. When it resumes from an interruption, it will start rendering the newest value from the beginning. This means React is only working on the UI that the user will actually need to see rendered, and never an old state.

[Timer queue and task queue are separated](https://github.com/facebook/react/blob/d20c3af9d11ea4a35bfc76cb44c15af9d42059c4/packages/scheduler/src/forks/Scheduler.js#L308), and when the time expires, it will be executed from the timer queue to the task queue.
It is sorted according to priority.
image

They increase responsiveness by [giving priority to sync, input, general, and idle](https://github.com/facebook/react/blob/d20c3af9d11ea4a35bfc76cb44c15af9d42059c4/packages/react-reconciler/src/ReactFiberLane.new.js#L131).

[Automatic batching](https://github.com/reactwg/react-18/discussions/21) is also possible when events and timeouts occur, which improves performance.
You can also do [batching](https://github.com/wilsonpage/fastdom) like read, layout, paint..etc when accessing dom other than virtual dom rendering.

More context
- https://github.com/facebook/react/pull/18796
- React fiber uses the [Left-child right-sibling binary tree](https://en.wikipedia.org/wiki/Left-child_right-sibling_binary_tree)
- https://engineering.fb.com/2019/04/22/developer-tools/isinputpending-api/

**Vue's Compiler-Informed virtual dom & Glimmer VM**

Vue provides a way to [compile components](https://vuejs.org/guide/extras/rendering-mechanism.html#compiler-informed-virtual-dom).
Use strategies such as static hoisting, patch flags, tree flattening.

Ember's glimmer generates a kind of [bytecode](https://engineering.linkedin.com/blog/2017/12/the-glimmer-binary-experience).
image
image

More context
- [Vue3 source code analysis (5): Patch algorithm](https://segmentfault.com/a/1190000040097158)
- [Compilation optimization of vue3: block tree and patchflags](https://cdmana.com/2022/116/202204261710474691.html)
- [Glimmer: Blazing Fast Rendering for Ember.js, Part 1](https://engineering.linkedin.com/blog/2017/03/glimmer--blazing-fast-rendering-for-ember-js--part-1)
- [Glimmer: Blazing Fast Rendering for Ember.js, Part 2](https://engineering.linkedin.com/blog/2017/06/glimmer--blazing-fast-rendering-for-ember-js--part-2)

**Solid's reactivity & state**

[Fine-grained reactivity](https://www.solidjs.com/guides/reactivity) can lead to minimal state changes.

image

An example of a neat library is [recoil](https://recoiljs.org/).
image

Another way to deal with a complex state without the reactive, is [statecharts](https://statecharts.dev/). (Famous library: [xstate](https://github.com/statelyai/xstate))

More context
-  [A Hands-on Introduction to Fine-Grained Reactivity](https://dev.to/ryansolid/a-hands-on-introduction-to-fine-grained-reactivity-3ndf)
- [SolidJS: Reactivity to Rendering](https://indepth.dev/posts/1289/solidjs-reactivity-to-rendering)
- [Integration with External State Systems](https://vuejs.org/guide/extras/reactivity-in-depth.html#integration-with-external-state-systems)
- [The fundamental principles behind MobX](https://hackernoon.com/the-fundamental-principles-behind-mobx-7a725f71f3e8)
- [#MobX or @recoiljs ?](https://twitter.com/mweststrate/status/1261369870152871937)
- [re-frame](https://day8.github.io/re-frame/a-loop/)

**Mikado's pooling**

Mikado has a [fairly unique pooling system](https://github.com/nextapps-de/mikado#concept-of-shared-pools).

image

**Style**

Freestyler caches styles according to  [cardinality](https://github.com/streamich/freestyler/blob/master/docs/en/how-it-works.md), 
Firefox has a [prototype cache](https://github.com/mozilla/gecko-dev/blob/master/dom/xul/nsXULPrototypeCache.h)

The most advanced aspect of the web in terms of style and theme is how it manages.
In constructing a design system[[1](https://www.bornfight.com/blog/design-systems-design-tokens-atomic-design-the-art-of-scaling-design/), [2](https://www.uxpin.com/studio/blog/design-systems-vs-pattern-libraries-vs-style-guides-whats-difference/), [3](https://www.justinmind.com/ui-design)] like [fast](https://www.fast.design/docs/category/creating-design-systems/)
- [Headless UI](https://medium.com/cstech/headless-ui-components-creating-re-usable-logic-without-thinking-about-design-69ac9fad6400): It is a component [without a style, providing reusable logic](https://open-ui.org/). https://github.com/linebender/druid/issues/2160
- [Design Token](https://spectrum.adobe.com/page/design-tokens/): If you use [on-demand atomic CSS](https://antfu.me/posts/reimagine-atomic-css) and can be adjusted by using the rules to adjust the size and brightness ..etc.

**Others**
- Event: Use the [event delegation](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_delegation) to reduce the number of listeners to improve performance. Most JS libraries do this automatically.
- [Litho](https://fblitho.com/): Asynchronous layout, Flatter view hierarchies, Fine-grained recycling UI framework for Android
